librz/bin/format/le/le.c

```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
+ * possible I/O scenarios:
+ *
+ *   A. inside (1)
+ *   B. on the boundary (1)-(2)
+ *   C. inside (2)
+ *   D. on the boundary (2)-(3)
+ *   E. touching (1)-(2)-(3) -- possible when (2) is very small and is covered by I/O
+ *
+ * Note that \param data_vaddr is required to be inside \param io_page, so it's impossible
+ * for I/O to cross the lower boundary of (1) or happen wholly to the right of (2).
+ *
+ * The function works by iterating maps left to right for as long as there's hope of finding
+ * intersections with the I/O interval. When patching fixups \param data_len never exceeds
+ * 6 bytes (for 16:32 fixups), but the algorithm should work for any \param data_len.
+ **/
+static bool page_io(rz_bin_le_obj_t *bin, LE_page *io_page,
+	ut32 data_vaddr, ut8 *data, ut32 data_len, bool read) {
+
+	for (ut32 mi = io_page->le_map_num - 1; mi < rz_vector_len(bin->le_maps); ++mi) {
+		LE_map *m = rz_vector_index_ptr(bin->le_maps, mi);
+		if (m->obj_num != io_page->obj_num) {
+			return true; // the map belonging to another object reached, stop
+		}
+
+		ut32 vfile_beg = m->vaddr;
+		ut32 vfile_end = m->vaddr + m->size;
+		ut32 vdata_beg = data_vaddr;
+		ut32 vdata_end = data_vaddr + data_len;
+		if (vdata_end <= vfile_beg) {
+			return true; // no further intersections possible, stop
+		}
+		if (vfile_end <= vdata_beg) {
+			continue; // no intersection yet, try next map
+		}
+
+		ut32 vbeg = RZ_MAX(vfile_beg, vdata_beg);
+		ut32 vend = RZ_MIN(vfile_end, vdata_end);
+		ut32 len = vend - vbeg;
+
+		RzBuffer *vfile_buf = m->is_physical ? bin->buf_patched : m->vfile_buf;
+		ut64 paddr = m->paddr + vbeg - vfile_beg;
+		ut8 *buf = data + vbeg - vdata_beg;
+		if (!vfile_buf) {
+			// likely vfiles haven't been created correctly
+			RZ_LOG_ERROR("LE: attempted %s %d byte(s) at 0x%" PFMT64x " of map %s "
+				     "with no buffer.\n",
+				read ? "reading" : "writing", len, paddr,
+				m->vfile_name ? m->vfile_name : "NULL");
+			rz_return_val_if_reached(false);
+		}
+		bool good;
+		if (read) {
+			good = rz_buf_read_at(vfile_buf, paddr, buf, len) == len;
+		} else {
+			good = rz_buf_write_at(vfile_buf, paddr, buf, len) == len;
+		}
+		if (!good) {
+			// likely data_vaddr is outside page_io, misused this function
+			RZ_LOG_ERROR("LE: error %s vfile, %d byte(s) at 0x%" PFMT64x ".\n",
+				read ? "reading" : "writing", len, paddr);
+			rz_return_val_if_reached(false);
+		}
+	}
+	return true;
+}
+
+static bool page_read(rz_bin_le_obj_t *bin, LE_page *io_page, ut32 vaddr, ut8 *buf, ut32 len) {
+	return page_io(bin, io_page, vaddr, buf, len, true);
+}
+
+static bool page_write(rz_bin_le_obj_t *bin, LE_page *io_page, ut32 vaddr, ut8 *buf, ut32 len) {
+	return page_io(bin, io_page, vaddr, buf, len, false);
+}
+
+static ut32 le_reloc_vaddr(rz_bin_le_obj_t *bin, LE_reloc *reloc) {
+	return bin->le_pages[reloc->src_page].vaddr + reloc->src_off;
+}
+
+static ut32 le_reloc_size(LE_reloc *reloc) {
+	static const ut8 szmap[] = {
```

Comment:
Probably better to extract to the top level

Reply:
I don't export exotic relocs from rz_bin_le_get_relocs, for example 6 byte 16:32 reloc is converted to a pair of a 32 bit reloc and a 16 bit selector reloc. I would need to add a new reloc type to do that properly, and I am not up to the task. So I judged it's an LE internal detail. I think it should not be exported.